Autogenerate images of parts of the viewer

TimMonko opened the pull request:
# References and Motivation

With 0.6.0 on the way, there have been _many_ changes to the UI from 0.5.6, especially when considering the years of napari development. Much of these changes have been done by me, and in recognition of that, I would like to update the current static images used throughout the docs with dynamically generated ones, so that any future UI changes do not require thorough updates of the docs. This PR was originally conceived slightly before, and then [discussed at 2025-03-13/14 docs meeting](https://hackmd.io/10DKrIMhREmk9sa_QRirlg#2025-03-1314); the docs meeting concluded that this PR should be made quickly, merged, and then checked for any flakiness (since it requires a merge) or appearance issues in the auto-generation. Then, we can make specific updates to the script (or new scripts) in order to improve visualization for various places in the docs. These are presently designed as proof-of-principle and change things as needed. Conceivably, this can also replace images where we hide building up a napari viewer instance, only to take an nbscreenshot. I believe this script method is considerably faster, but does require a bit less intuitive of a setup compared to the screenshot method. Overall, this takes less than 3 seconds to generate over 20 images. I'm aware that many of these are not currently needed in the docs, but may provide inspiration and design principles, even for the autogeneration, moving forward. We'll clean up what exactly is autogenerated in follow-up PRs. In follow ups, we can also consider adding unique stylings to Qt elements on the fly, such as making a certain button of interest have a yellow border. 

Useful current places:
1. [viewer tutorial](https://napari.org/dev/tutorials/fundamentals/viewer.html#viewer-buttons) buttons, popups (e.g. we just added grid popup and the buttons below are already out of date),, anything with console, etc. quite a few static images in here
2. [layers tutorial](https://napari.org/dev/guides/layers.html)
3. [quick start](https://napari.org/dev/tutorials/fundamentals/quick_start.html). console related images
4. [various buttons and images in how to each layer](https://napari.org/dev/howtos/layers/image.html)

And in the future, because we can more quickly generate images, we can conceivably use more focused images throughout the docs. 

This will require doing something with doc merges as force-orphan to prevent build up of repo size with changing binaries, though nicely, these images are so far quite small in size (comparatively)

# Description

In one single file, our goal is to autogenerate various elements of the UI; these are either initially gotten with `viewer.screenshot` of the current viewer state, or in follow-ups as specific pixmaps of the widgets (or a defined group of widgets). These pixmaps are defined by dictionaries at the top of the file, making it easier to find and contribute to an image of interest. It also serves as models for 'set up' of different element types (popups, menus, other states)

These include:
1. single widget elements always visible in the UI 
2. menus, and specific submenus
3. triggered popup widgets, like grid, roll, ndisplay
4. groups of widgets. these are my favorite, but will take the most thought to define because it allows us to dynamically 'crop' the viewer to areas of interest. Lovely! 

The current implementation prioritizes speed, but may need to add some QTimer / sleep events for CI, not sure. 

# Autogenerated image examples

## Viewer and folder structure
![image](https://github.com/user-attachments/assets/dc4b05f3-f485-4012-8d94-845f358f34c2)

## Single widgets
![image](https://github.com/user-attachments/assets/5a4fddad-410c-49f8-b3a1-e3e8ca23477c)

## Menus
![image](https://github.com/user-attachments/assets/c86426c9-6979-42ce-80ce-6cbdd3a03bf5)

## Popups
![image](https://github.com/user-attachments/assets/cc2611a9-95c2-4c8f-a35f-f81ad2285518)

## Combined groups of widgets
![image](https://github.com/user-attachments/assets/8a0ce0bf-5c37-407a-a6e4-571c19785c39)


# Original PR Description

This is _really_ a draft of trying to autogenerate images that we can use throughout the docs. I think this would _also_ speed up doc generation in certain areas with reusable doc images, so that nbscreenshot would not need to be as frequent. Everything is getting saved into `images/_autogenerated` and could be used elsewhere.

1. Tries to autogenerate grabbing and saving many parts of the viewer using the attributes of the viewer

Needs:
1. Popups are not working right, sometimes they trigger, often its the dims slider popup. I probably need to just directly call the button, but I'm having a tough time figuring out where the buttons live. 
6. I think the timing of things has me really confused. I think QTimer is necessary to wait for things to popup, but I can't figure out why these popups aren't working
7. Also sometimes the console popups and saves as a screenshot, other times its just the thin dock area prior to the console popup. 

Examples of what _should_ appear in autogenerated if CI does it the same was as locally
![image](https://github.com/user-attachments/assets/a9e3be59-bd61-481e-a046-62251177f5a8)

TimMonko committed:
autogeneration-preliminary-work

TimMonko committed:
clean up

TimMonko committed:
barely working popups skeleton

TimMonko committed:
im lost this popup sometimes works but usually doesn't

TimMonko commented:
Adding a few links from the community meeting today:
https://simonwillison.net/2022/Oct/14/automating-screenshots/
https://simonwillison.net/2022/Mar/10/shot-scraper/
https://github.com/napari/napari/pull/7678
and look into qt_bot wait

TimMonko committed:
cleaner working popups

TimMonko committed:
gui includes both viewer and popups -- but popups sometimes don't work

TimMonko committed:
seems to work less but code is better

melissawm commented:
Hi @TimMonko - this is the PR I mentioned on zulip: #308 

That one is more focused on videos, so might not be helpful but just in case here it is. Cheers!

Czaki commented on docs/_scripts/autogenerate_popups.py:
```diff
@@ -0,0 +1,92 @@
+            "grid_popup", 
+            lambda: cleanup(viewer)
+        )
+    ))
```
This is not a good solution. We need to find a way to define a list of popups to iterate over it. 
Maybe global list and just pass the index?

TimMonko replied:
can you look strictly at autogenerate-gui? I've put the most recent popup stuff in there. It's flaky, where sometimes it works and sometimes it doesn't, I've tried to include closing popups there, but yes, trying to work towards creating a list. Sorry I didn't remove this file 😬

Czaki replied:
in 2-3 hours

TimMonko replied:
of course! no issue, I'm actually going to plug along on it for about another hour, so you should receive a better file by then

TimMonko replied:
your review did at least confirm some of the direction I thought it needed to go

Czaki commented on docs/_scripts/autogenerate_popups.py:
```diff
@@ -0,0 +1,92 @@
+    popup = None
+    for widget in QApplication.topLevelWidgets():
+        if isinstance(widget, QtPopup):
+            popup = widget
```
The previous popup is not closed before your code triggers creation of the next popup, so the first popup is captured again. We either need to use proper class in place of `QtPopup` or use `objectName()` method (with earlier setting it).

TimMonko committed:
remove popups file -- outdated

TimMonko committed:
clean up widget and menu popup capturing

TimMonko committed:
got popups working consistently

TimMonko committed:
add more popups, clean up code more

TimMonko committed:
speed up slightly

TimMonko commented:
@Czaki 
Everything is organized in one file now, and the script is not flaky anymore. 
QTimer actually seems to create more problems than fewer problems for popups, compared to menus. Perhaps I was implementing incorrectly --though I was using same logic as menus, it would be delayed asynchronously. Strangely, lowering Qtimer for capturing menus to 50ms resulting in incomplete generation of popups. The async part of this is so confusing.

1. Creates images of many widgets
2. Creates images of many menu items
3. Creates images of popups with prep triggers. 

We can add / change more and modify callbacks in various ways, but as @willingc we can start with a merged script and go from there.

I hope it's not flaky on anyone else's machine, or CI.

TimMonko committed:
lengthen menu QTimer so as not to cause issue with popup generation

willingc commented on docs/_scripts/autogenerate_gui_images.py:
```diff
@@ -0,0 +1,238 @@
+    viewer.window._qt_window.resize(1000, 800)
+    viewer.window._qt_window.setStyleSheet(get_stylesheet("dark"))
+    
+        # Ensure window is active
```
```suggestion
    # Ensure window is active
```

willingc reviewed (approved):
Let's merge this as is. I would like to do a pass over the script and change some of the print statements to logging (it will benefit us when running in CI).

TimMonko commented:
Sounds good, the prints were really just for debugging lol. I take them out as things start to work. Definitely anything with prints from me is a 'draft'. Can still merge though if you like

We should be able to see if CI autogenerates correctly

willingc commented:
I would like us to merge and iterate on it.

melissawm commented:
Can we add a comment in the script to indicate which images it is generating?

TimMonko committed:
Merge branch 'main' into autogenerate-gui

TimMonko committed:
Refactor autogenerate_images function and modularize widget, menu, and popup handling

TimMonko committed:
add mouse over for status bar

TimMonko committed:
improve docstrings

TimMonko committed:
Add functionality to capture viewer regions

TimMonko commented:
@melissawm I added much more commenting to try to clarify what is being saved and I _hope_ the code organization is sufficient to be self explanatory.  _However_, if what you mean is to specifically say which images for the docs this is generating, I'm not sure that's a good idea (will be hard to maintain and keep up to date). Please let me know if I haven't accomplished the goal.

I've also added various interactions with the canvas to change the display of various widgets -- we can always be very specific if we have things in mind to demonstrate. At this point, I'm very happy with this implementation, and want us to merge this to check it works on CI, and then we can look to specifically define images of interest that can be used in the docs. 

I also added a preliminary, rough implementation of capturing _regions_ of the viewer as defined by the widgets. In the long run, this will be more maintainable than defining the region with our own coordinates (since inevitably these coordinates could move around). Currently, it just crops to the bounds of the defined widgets, for example:

Viewer buttons + Console:
![console_and_buttons](https://github.com/user-attachments/assets/8f11bcaf-862d-4a56-ae61-db85cda4d99e)

Layer list and layer controls 
![layer_list_and_controls](https://github.com/user-attachments/assets/5ac0713f-e76b-49d1-9945-ad88d2d55b55)

melissawm commented:
Ok - Maybe I just need more context on what we're accomplishing. Reading this is seems like we are generating one image for the empty viewer, one for the cells3d sample and another for cells3d with the console. My question is - are we supposed to use this script in the future to include all other images? If not, why are we only using those 3?

I don't want to block this though, feel free to merge and iterate on it later!

TimMonko committed:
Merge branch 'main' into autogenerate-gui

TimMonko commented:
@melissawm updated the PR description to fully encapsulate the goals of this PR and the description of every image that is taken.

melissawm commented:
Ok - having now taken a better look, I think this is great and we should merge as-is. This is at the very least an excellent starting point for generating images from widgets and GUI components. It's not running right now but we can figure out that piece in a follow-up. I'm taking the liberty to merge since this doesn't change anything for the docs right now but it's an important piece for whatever we do next.

Thanks again @TimMonko !

melissawm committed:
Merge branch 'main' into autogenerate-gui
